codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/QueryShapeSerVisitor.java

```diff
@@ -97,29 +94,43 @@ protected void serializeMap(GenerationContext context, MapShape shape) {
             // Use the @xmlName trait if present on the member, otherwise use "key".
             MemberShape keyMember = shape.getKey();
             Shape keyTarget = model.expectShape(keyMember.getTarget());
-            String keyName = keyMember.getTrait(XmlNameTrait.class)
-                    .map(XmlNameTrait::getValue)
-                    .orElse("key");
-            writer.write("const keyLoc: string = \"entry.$L.\" + counter;", keyName);
-            writer.write("entries[keyLoc] = $L;", keyTarget.accept(getMemberVisitor("key")));
+            String keyName = getMemberSerializedLocationName(keyMember, "key");
+            writer.write("entries[\"entry.$L.\" + counter] = $L;", keyName, keyTarget.accept(getMemberVisitor("key")));
 
 
             // Prepare the value's entry.
             // Use the @xmlName trait if present on the member, otherwise use "value".
             MemberShape valueMember = shape.getValue();
             Shape valueTarget = model.expectShape(valueMember.getTarget());
-            String valueName = valueMember.getTrait(XmlNameTrait.class)
-                    .map(XmlNameTrait::getValue)
-                    .orElse("value");
-            writer.write("const valueLoc: string = \"entry.$L.\" + counter;", valueName);
-            writer.write("entries[valueLoc] = $L;", valueTarget.accept(getMemberVisitor("input[key]")));
-
+            String valueName = getMemberSerializedLocationName(valueMember, "value");
+            serializeUnnamedMemberEntryList(context, valueTarget, "input[key]", "entry." + valueName);
             writer.write("counter++;");
         });
 
         writer.write("return entries;");
     }
 
+    private void serializeUnnamedMemberEntryList(
+            GenerationContext context,
+            Shape target,
+            String inputLocation,
+            String entryWrapper
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+
+        QueryMemberSerVisitor inputVisitor = getMemberVisitor(inputLocation);
+        // Map entries that supply entry lists need to have them joined properly.
+        if (inputVisitor.visitSuppliesEntryList(target)) {
+            writer.write("const valueLoc: string = \"$L.\" + counter + \".\";", entryWrapper);
```

Comment:
I think this kind of concatenation can be replaced with TypeScript template strings:  https://howtodoinjava.com/typescript/template-strings/ (remember, you'll need to use double `$` to escape the CodeWriter templating).